gcs/src/gcs_gcomm.cpp

```diff
@@ -230,6 +230,29 @@ class GCommConn : public Consumer, public Toplay
 
     ~GCommConn()
     {
+        // We cannot call gcs_close() to perform cleanup if there is an
+        // exception. So, we need to explicity free the objects in the
+        // destructor.
+        if (tp_ != nullptr)
+        {
+            delete tp_;
+        }
+
+        mutex_.lock();
+        if (!terminated_)
+        {
+            {
+                gcomm::Critical<Protonet> crit(*net_);
+                log_info << "gcomm: terminating thread";
+                terminated_ = true;
+                notify();
+            }
+            mutex_.unlock();
+            log_info << "gcomm: joining thread";
+            gu_thread_join(thd_, 0);
+            mutex_.lock();
+        }
+        mutex_.unlock();
```

Comment:
LGTM
I don't know if it would be better, probably yes because of getting rid of unnecessary lock/unlock cycle, but consider this setting unlock_needed = true before 'if', then unlock_needed = false in 'if' after unlocking and then if(unlock_needed) mutex._unlock() after 'if'

Reply:
Done